static/js/drafts.js

```diff
@@ -143,13 +155,53 @@ exports.restore_draft = function (draft_id) {
     }
 
     overlays.close_overlay("drafts");
-    compose_fade.clear_compose();
-    if (draft.type === "stream" && draft.stream === "") {
-        draft_copy.subject = "";
+    load_restored_draft(draft_id);
+};
+
+exports.restore_on_narrow = function (operators) {
+    var filter = new Filter(operators);
+    var draft_list = Object.keys(draft_model.get());
+    var restore;
+    if (filter.has_operator('pm-with')) {
+        if (operators.length > 1) {
+            return;
+        }
+        draft_list = _.filter(draft_list, function (draft_id) {
+            if (draft_model.getDraft(draft_id).type === 'stream') {
+                return false;
+            }
+
+            var recip = draft_model.getDraft(draft_id).private_message_recipient;
+            recip = recip.replace(/\s/g, '').toLowerCase().split(',').sort().join(',');
+            var op_recip = operators[0].operand.split(',').sort().join(',');
+
+            return recip === op_recip;
+        });
+        restore = draft_list.pop();
+    } else if (filter.has_operator('stream') && filter.has_operator('topic')) {
+        if (operators.length > 2) {
+            return;
+        }
+        draft_list = _.filter(draft_list, function (draft_id) {
+            if (draft_model.getDraft(draft_id).type === 'private') {
+                return false;
+            }
+
+            return _.all(operators, function (op) {
+                if (op.operator === 'stream') {
+                    return op.operand === draft_model.getDraft(draft_id).stream;
+                }
+                return op.operand === draft_model.getDraft(draft_id).subject;
+            });
+        });
+        restore = draft_list.pop();
     }
-    compose_actions.start(draft_copy.type, draft_copy);
-    compose_ui.autosize_textarea();
-    $("#new_message_content").data("draft-id", draft_id);
+
+    if (!restore) {
+        return;
+    }
+
+    load_restored_draft(restore);
 };
```

Comment:
For testing purposes, I'd consider making this function be like this:

~~~ js
exports.restore_on_narrow = function (operators) {
   var draft = exports.draft_matching_narrow(operators);
   if (draft) {
        load_restored_draft(draft);
   }
}
~~~

And then move most of the logic that is immutable into `draft_matching_narrows`.